File: tests/test_exercise_context.js
/* Run after export_app.py: node tests/test_exercise_context.js [baseline-content.json] */
'use strict';
const assert = require('node:assert/strict');
const fs = require('node:fs');
const path = require('node:path');
const vm = require('node:vm');
const root = path.resolve(__dirname, '..');
const content = fs.readFileSync(path.join(root, 'data/app/content.json'), 'utf8');
const core = fs.readFileSync(path.join(root, 'app/js/00_core.js'), 'utf8');
const learn = fs.readFileSync(path.join(root, 'app/js/40_learn.js'), 'utf8').replace(
  'return { render, open, rerenderBlock, refreshNotes, state: st };',
  'return { render, open, rerenderBlock, refreshNotes, state: st, testRender: renderBlock, testVisible: visible };'
);
function contextFor(json) {
  const ctx = vm.createContext({ console, setTimeout, clearTimeout,
    document: {
      getElementById: () => ({ textContent: json }), addEventListener() {},
      createElement: () => ({ set innerHTML(value) { this.content = { firstElementChild: { outerHTML: value } }; } })
    }, window: { addEventListener() {} }
  });
  vm.runInContext(core, ctx);
  const historyPath = path.join(root, 'app/js/05_history.js');
  if (fs.existsSync(historyPath)) vm.runInContext(fs.readFileSync(historyPath, 'utf8'), ctx);
  return ctx;
}
const ctx = contextFor(content);
vm.runInContext(learn, ctx);
vm.runInContext(`
  const unit = SUBJ.science.byCode.R13, block = unit.content[2];
  const question = unit.items.find(i => i.ci === 2 && i.n === 1);
  if (question.key !== 'h5:science/R13#2:1') throw Error('persistent item key changed');
  if (!question.context.includes('<svg')) throw Error('card lost its shared circuit');
  if (question.q.includes('<svg')) throw Error('shared circuit duplicated in question');
  const card = studyCard(question, 'science');
  if (!card.context.includes('<svg') || !card.label.includes('(1)')) throw Error('card adapter lost context or numbering');
  ['all', 'ex'].forEach(tab => {
    Learn.state.tab = tab;
    if (!Learn.testVisible(block)) throw Error('exercise hidden');
    const rendered = Learn.testRender(unit, block).outerHTML;
    const originalCount = (block.context.match(/<svg/g) || []).length;
    if ((rendered.match(/<svg/g) || []).length !== originalCount) throw Error('shared circuit absent or duplicated');
  });
  Learn.state.tab = 'ex';
  if (!Learn.testVisible({t:'q'}) || !Learn.testVisible({t:'data'})) throw Error('raw question or data hidden in exercises');
  Learn.state.tab = 'points';
  if (Learn.testVisible({t:'q'})) throw Error('raw question incorrectly shown as a knowledge point');
  const answerUnit = SUBJ.science.units.find(u => u.code === 'M1' && u.title.includes('解答'));
  if (answerUnit && answerUnit.content.filter(c => c.t === 'a').some(c => !(c.answerKeys || []).length)) throw Error('a separated mock answer lost its question exposure mapping');
`, ctx);
const baseline = process.argv[2];
if (baseline) {
  const previous = contextFor(fs.readFileSync(baseline, 'utf8'));
  const keys = vm.runInContext('EDITIONS.flatMap(e => e.subjects.flatMap(s => s.units.flatMap(u => u.items.map(it => it.key))))', previous);
  const newKeys = new Set(vm.runInContext('EDITIONS.flatMap(e => e.subjects.flatMap(s => s.units.flatMap(u => u.items.map(it => it.key))))', ctx));
  assert(keys.every(key => newKeys.has(key)), 'an existing grade key disappeared');
  const oldShape = vm.runInContext('EDITIONS.flatMap(e => e.subjects.flatMap(s => s.units.map(u => [u.ukey, u.content.length])))', previous);
  const newShape = vm.runInContext('EDITIONS.flatMap(e => e.subjects.flatMap(s => s.units.map(u => [u.ukey, u.content.length])))', ctx);
  const shapeMap = new Map(newShape);
  assert(oldShape.every(([ukey, count]) => shapeMap.has(ukey) && shapeMap.get(ukey) >= count), 'an existing unit or block position disappeared');
  console.log('Preserved ' + keys.length + ' existing item keys and the original units/block slots (additional content allowed).');
}
console.log('Question contexts, mock visibility, and card adapters passed.');

File: tests/test_exercise_context.py
"""Run with python -m unittest discover -s tests -p 'test_exercise_context.py'."""
import sys
import unittest
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT / 'build'))
from export_json import split_items, split_item_block, parse
from export_app import build_subject, linked_mock_questions
import editions as ED


class ExerciseContextTests(unittest.TestCase):
    def test_shared_table_figure_and_prose_survive_split(self):
        prefix = '資料A：観察の条件\n\n| 実験 | 結果 |\n|:--|:--|\n| A | 変化した |\n\n[[svg:circuit_sp]]'
        block = split_item_block(prefix + '\n\n(1) 何を比べるか。\n(2) 理由を答えよ。')
        self.assertEqual(block['context_md'], prefix)
        self.assertEqual([i['no'] for i in block['items']], [1, 2])
        self.assertNotIn('資料A', block['items'][0]['text'])
        self.assertEqual(block['items'], split_items(prefix + '\n\n(1) 何を比べるか。\n(2) 理由を答えよ。'))

    def test_existing_inline_numbers_keep_their_values(self):
        items = split_items('8. 前半　9. 中盤　10. 後半')
        self.assertEqual([i['no'] for i in items], [8, 9, 10])
        self.assertIsNone(split_item_block('1. A\n3. B'))

    def test_h5_circuit_and_history_context_present(self):
        ed = ED.by_id('h5')
        science = build_subject('science', 'S', '理科', '理', ed)
        social = build_subject('social', 'H', '社会', '社', ed)
        circuit = next(u for u in science['units'] if u['code'] == 'R13')['content'][2]
        self.assertIn('<svg', circuit['context'])
        self.assertEqual([i['n'] for i in circuit['items']], list(range(1, 8)))
        self.assertTrue(all('<svg' not in i['q'] for i in circuit['items']))
        history = next(u for u in social['units'] if u['code'] == 'R2')
        context = '\n'.join(c.get('context', '') for c in history['content'])
        self.assertIn('望月', context)
        self.assertIn('資料', context)

    def test_mock_questions_link_only_to_their_own_answer_set(self):
        ed = ED.by_id('h5')
        for sid, key, name, short, count in [('science', 'S', '理科', '理', 5), ('social', 'H', '社会', '社', 6)]:
            sub = build_subject(sid, key, name, short, ed)
            for code in ['M1', 'M2']:
                mock = next(u for u in sub['units'] if u['code'] == code)
                questions = [c for c in mock['content'] if c['t'] == 'ex']
                self.assertEqual(len(questions), count)
                self.assertTrue(all(c['a'] and c['contextTitle'].startswith('大問') for c in questions))
                self.assertTrue(all(c['t'] != 'q' for c in mock['content']))
                # Original answer units remain present; note/ink indices cannot shift.
                answers = [u for u in sub['units'] if u['code'] == code and '解答' in u['title']]
                self.assertEqual(len(answers), 1)

    def test_ambiguous_answer_set_is_not_guessed(self):
        sec = {'file': 'mock.md', 'path': ['章', 'M1　模試', '大問1'], 'heading': '大問1', 'blocks': [{'type': 'q', 'title': '問題', 'md': '1. A\n2. B'}]}
        ans = {'file': 'mock.md', 'path': ['章', 'M1　解答'], 'heading': 'M1　解答', 'blocks': [{'type': 'a', 'title': '大問1', 'md': '1. A\n2. B'}] * 2}
        self.assertEqual(linked_mock_questions({'sections': [sec, ans]}), {})


if __name__ == '__main__':
    unittest.main()
